### DISCUSSION TITLE
How do we communicate a warning/notice of eia early release data?

### DISCUSSION THREAD

**Discussion by cmgosnell** [catalyst-cooperative/pudl#1799] [Help Me!] [answered]
If we are to integrate early release data (draft releases of full years) into pudl, how do we want to communicate the fact that the new year of data is early release? Some options:

- [ ] add a logger/warning (easy but will not help datasette users)
- [ ] add info into data set level metadata (this will show up in datasette web interface, but once - [ ] users pull the data down, they could very easily forget)
- [ ] add some flag in a/any table w/ early release data?

**Comment by cmgosnell** [comment, marked as answer]
There is a RIDICULOUSLY long-named column in the early release data (`early_release_data_june_2022_not_fully_edited_use_with_caution_do_not_aggregate_to_state_regional_or_national_totals`).

If we went with the latter option of adding a flag, I could imagine using this column or generating another column altogether. We could use this for the eia860m data as well. iirc we are not reeeeeally communicating that the newest 860 year is a monthly release. in part bc it is of annual-type data (i.e. `capacity_mw` not `net_generaiton_mwh`)

**Reply by zaneselvans** [reply]
I think if we're going to have a mix of "final" historical data and year-to-date updates (eia860m) / early release data, having a data source flag column in the monthly & annual data would be the clearest option. Maybe with values like `eia860m`, `eia860er` and `eia923er` Though I guess whether an attribute comes from the `eia923` or `eia860` gets kind of fuzzy for the tables that are built during normalization / harvesting. But we generally attribute the annual entity data to the `eia860` as it is now.

**Reply by cmgosnell** [reply]
I agree with this overall! I think something like a `datasource` column (with `eia923`, `eia923er`, `eia860`, `eia860er`, `eia860m`, etc) makes sense to me.

For the standard data tables (the ones that we harvest entity data *from*), this table is very straightforward imo.


### On implementation
Where to store the info of whether a specific year is  Specific DOI's and sets of metadata for a particular year will necessary correspond to either the `er` version of an eia data-year. We could store this info in `pudl.metadata.sources` and access it through `DataSource.from_id()`. In that context, because we know we are never going to have multiple years of early release data, I would propose adding `early_release: int|None` into eia data sources where you could either give a year or None.

We could add it into the extract metadata but I don't think this would be preferable.

How/where should we generate this flag?
* In the extract step?
* In the transform step? (i think i prefer this one)

**Reply by zaneselvans** [reply]
I think the right place to capture and store this information is probably in the scraper (since it's information that's contained in the web page we're scraping -- and that's where we would look it up manually) and Zenodo archives. Otherwise nobody can tell whether the archived raw data is final or not, and we would have to be very diligent about (manually) updating the DataSource metadata every time something transitions from ER to final.  If it's captured in the scraper, then as soon as the published data transitions it would be reflected in the whole chain. The EIA specific `ExcelExtractor` classes can look up the per-year source metadata from the data package (`eia860` vs. `eia860m` vs `eia860er` etc.) and insert it into a column in the `process_raw()` method.

**Reply by cmgosnell** [reply]
I agree in practice that the scraper should be the place that a) knows and b) documents the `er`-ness. But in practice this would be challenging to implement with the current split of the scraper vs the archiver (which makes the metadata). If we ever combine them in the future this could be handled there np! For now, I think we can handle this by using the file names in the `file_map.csv` because EIA embeds `Early_Release` into the file names. This could be read within the `ExcelExtractor`. Because the `ExcelExtractor` knows both about the datastore and the file maps, this is a nice way imo to enable us to change this now using the filenames and swoop in later to change it in the same place without a big refactor.


One question @zaneselvans ... we have a `data_source` column that we add to `eia860` and `eia860m` tables and then harvest for the `generators_eia860` table. This is *just* naming but I am bristling at using this column to distinguish the early release data. Perhaps `source` is a fine name here but I have been thinking of this more of a `data_label` than a `data_source`. EIA923 is the source. We are attempting to alert users that this particular EIA923 record comes from the early release. `eia80m` is somewhat different imo but I think it could still fit under this idea of a `data_label` instead of a `data_source`. Thoughts?

**Reply by zaneselvans** [reply]
`data_label` seems too generic. That could be like... tags or keywords or something.

What are we really trying to convey here?  We have `DataSource` already defined in our little universe, but now it turns out there's a subdivision within `DataSource`, such that the same information is being provided in the same structure, but at a different level of maturity? and what is this new dimension? Data maturity? Data version? Release version?

What about `data_source_version`?

**Reply by zaneselvans** [reply]
Or if we wanted to try and enforce a uniform taxonomy across the board we could just have `data_maturity` or `data_revision`, not include the form, and instead have a controlled vocabulary that we define like...
* `year_to_date`
* `early_release`
* `final`

**Reply by cmgosnell** [reply]
I think I like `data_maturity`/ having a uniform taxonomy that does not actually include the `data_source`. I think these three options are good but I think we should add `monthly_release` for the eia860m. mostly because it's not reeeeally `year_to_date` like the 923 or cems YTD data is in that it is the monthly release of annual/static data.

**Reply by zaneselvans** [reply]
Hmm, to me `monthly_release` sounds much more like monthly data than annual data.

**Reply by zaneselvans** [reply]
But anyway, if we want to do a controlled vocab like this we should look at all the data sources we already know about and identify the existing universe of different kinds of maturity so we can make sure our categories can encompass them all clearly.
